src/Microsoft.AspNetCore.JsonPatch/Helpers/DictionaryWrapper.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.JsonPatch.Exceptions;
+
+namespace Microsoft.AspNetCore.JsonPatch.Helpers
+{
+    internal class DictionaryWrapper<TKey, TValue> : IDictionaryWrapper
+    {
+        private readonly IDictionary<TKey, TValue> _targetDictionary;
+        private readonly bool useDynamic;
+
+        public DictionaryWrapper(IDictionary<TKey, TValue> targetDictionary)
+        {
+            _targetDictionary = targetDictionary;
+            KeyType = typeof(TKey);
+            ValueType = typeof(TValue);
+
+            // For the case when target dictionary is an ExpandoObject (which implements IDictionary<string, object>)
+            useDynamic = (KeyType == typeof(string)) && (ValueType == typeof(object));
```

Comment:
As mentioned in my other comment, this code is not longer required as its being handled elsewhere, but your comment is valid though.